kubevela/kubevela
# Feat: Support SSH for retrieving terraform modules in private git repo

### Support SSH for retrieving terraform modules in private git repo

Enable support for terraform modules in private git repo through SSH. A kubernetes secret which holds the ssh private-key and known_hosts should be created and referenced in the component definition.

Terraform controller PR: https://github.com/kubevela/terraform-controller/pull/349

Feature for https://github.com/kubevela/terraform-controller/issues/292

I have:


### How has this code been tested


Tested in a local cluster. 
1. Create a kubernetes secret of type `kubernetes.io/ssh-auth` for the git ssh credentials & known-hosts
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: git-ssh
  namespace: vela-system
type: kubernetes.io/ssh-auth
stringData:
  ssh-privatekey: |
   <SSH Private Key> # the ssh private key used for authenticating git
  known_hosts: |
   <SSH known_hosts>  # use `ssh-keyscan github.com`  to generate known_hosts
```

2. Create a component definition
``` yaml
apiVersion: core.oam.dev/v1beta1
kind: ComponentDefinition
metadata:
  annotations:
    definition.oam.dev/description: Terraform module to provision a KMS key with alias
  labels:
    type: terraform
  name: aws-kms-key
  namespace: vela-system
spec:
  schematic:
    terraform:
      configuration: git@github.com:motilayo/terraform-aws-kms-key.git # use terraform configuration in private git repo to test
      gitCredentialsSecretReference: # reference to the secret created earlier
        name: git-ssh # secret name
        namespace: vela-system  # secret namespace
      type: remote
```

3. Create an application and verify
### Special notes for your reviewer
